src/bridge/cockpitdbussetup.c

```diff
@@ -0,0 +1,600 @@
+
+  gecos_dir_shell = g_hash_table_lookup (context->users, sp->sp_namp);
+  if (gecos_dir_shell && sp->sp_pwdp && strlen (sp->sp_pwdp) >= 4)
+    {
+      g_ptr_array_add (context->results,
+                       g_strdup_printf ("%s:%s:::%s", sp->sp_namp, sp->sp_pwdp, gecos_dir_shell));
+    }
+}
+
+static void
+setup_transfer_admin1 (GVariant *parameters,
+                       GDBusMethodInvocation *invocation)
+{
+  const gchar *groups[] = { "root", "wheel", NULL };
+  TransferAdmin1 *context = NULL;
+  const gchar *mechanism;
+  GVariant *prepared;
+  GVariant *result;
+
+  g_variant_get (parameters, "(&sv)", &mechanism, &prepared);
+
+  if (!g_str_equal (mechanism, "admin1"))
+    {
+      g_dbus_method_invocation_return_error (invocation, G_DBUS_ERROR, G_DBUS_ERROR_NOT_SUPPORTED,
+                                             _("Unsupported setup mechanism: %s"), mechanism);
+      goto out;
+    }
+  if (!g_variant_is_of_type (prepared, G_VARIANT_TYPE_STRING_ARRAY))
+    {
+      g_dbus_method_invocation_return_error (invocation, G_DBUS_ERROR, G_DBUS_ERROR_INVALID_ARGS,
+                                             _("Bad data passed for admin1 mechanism"));
+      goto out;
+    }
+
+  context = g_new0 (TransferAdmin1, 1);
+  context->groups = groups;
+  context->members = g_hash_table_new_full (g_str_hash, g_str_equal, g_free, NULL);
+  context->gids = g_hash_table_new (g_direct_hash, g_direct_equal);
+  context->users = g_hash_table_new_full (g_str_hash, g_str_equal, g_free, g_free);
+  context->results = g_ptr_array_new_with_free_func (g_free);
+
+  if (!fgetgrent_callback (filter_and_load_group, context) ||
+      !fgetpwent_callback (filter_and_load_user, context) ||
+      !fgetspent_callback (filter_and_add_results, context))
+    {
+      g_dbus_method_invocation_return_error (invocation, G_DBUS_ERROR, G_DBUS_ERROR_FAILED,
+                                             _("Couldn't load admin user data"));
+    }
+  else
+    {
+      result = g_variant_new_strv ((const gchar *const *)context->results->pdata, context->results->len);
+      g_dbus_method_invocation_return_value (invocation, g_variant_new ("(v)", result));
+    }
+
+out:
+  if (context)
+    {
+      g_hash_table_unref (context->members);
+      g_hash_table_unref (context->gids);
+      g_hash_table_unref (context->users);
+      g_ptr_array_free (context->results, TRUE);
+      g_free (context);
+    }
+  g_variant_unref (prepared);
+}
+
+typedef struct {
+  GDBusMethodInvocation *invocation;
+  GBytes *chpasswd;
+} CommitAdmin1;
+
+static void
+commit_admin1_free (gpointer data)
+{
+  CommitAdmin1 *context = data;
+  g_bytes_unref (context->chpasswd);
+  g_object_unref (context->invocation);
+  g_free (context);
+}
+
+static gboolean
+check_pipe_exit_status (CockpitPipe *pipe,
+                        const gchar *problem,
+                        const gchar *prefix)
+{
+  GError *error = NULL;
+  gint status;
+
+  if (problem == NULL ||
```

Comment:
It may not be possible but Is there a way to log better error messages here, right now all i get is child process exited with code 1.

Running the same command on the command line gives me a real error about groups.